src/infra/core/src/remote_alias_resolver_impl.rs

```diff
@@ -0,0 +1,305 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::str::FromStr;
+use std::sync::Arc;
+
+use auth::OdfServerAccessTokenResolver;
+use dill::*;
+use internal_error::{InternalError, ResultIntoInternal};
+use kamu_core::*;
+use opendatafabric::{self as odf, DatasetRefRemote};
+use url::Url;
+
+use crate::UrlExt;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct RemoteAliasResolverImpl {
+    remote_repo_reg: Arc<dyn RemoteRepositoryRegistry>,
+    access_token_resolver: Arc<dyn OdfServerAccessTokenResolver>,
+    dataset_repo: Arc<dyn DatasetRepository>,
+    remote_alias_reg: Arc<dyn RemoteAliasesRegistry>,
+}
+
+#[component(pub)]
+#[interface(dyn RemoteAliasResolver)]
+impl RemoteAliasResolverImpl {
+    pub fn new(
+        remote_repo_reg: Arc<dyn RemoteRepositoryRegistry>,
+        access_token_resolver: Arc<dyn OdfServerAccessTokenResolver>,
+        dataset_repo: Arc<dyn DatasetRepository>,
+        remote_alias_reg: Arc<dyn RemoteAliasesRegistry>,
+    ) -> Self {
+        Self {
+            remote_repo_reg,
+            access_token_resolver,
+            dataset_repo,
+            remote_alias_reg,
+        }
+    }
+
+    async fn fetch_remote_alias(
+        &self,
+        local_handle: &odf::DatasetHandle,
+        remote_alias_kind: RemoteAliasKind,
+    ) -> Result<Option<odf::DatasetRefRemote>, ResolveAliasError> {
+        let remote_aliases = self
+            .remote_alias_reg
+            .get_remote_aliases(&local_handle.as_local_ref())
+            .await
+            .int_err()?;
+
+        let mut push_aliases: Vec<_> = remote_aliases.get_by_kind(remote_alias_kind).collect();
+
+        match push_aliases.len() {
+            0 => Ok(None),
+            1 => Ok(Some(push_aliases.remove(0).clone())),
+            _ => Err(ResolveAliasError::AmbiguousAlias),
+        }
+    }
+
+    fn combine_remote_alias(
+        &self,
+        repo_url: &Url,
+        account_name_maybe: Option<odf::AccountName>,
+        dataset_name: &odf::DatasetName,
+    ) -> Result<odf::DatasetRefRemote, InternalError> {
+        let mut res_url = repo_url.clone().as_odf_protocol().int_err()?;
+        if let Some(account_name) = account_name_maybe {
+            res_url.path_segments_mut().unwrap().push(&account_name);
```

Comment:
`res_url.path_segments_mut().unwrap()` can be extracted to a local mut ref to simplify code a bit